### PR TITLE
AirfRANS: lr=7e-4 + grad-clip=1.5 — fill clip threshold gap

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Grad-clip=1.0 at lr=7e-4 achieved the best AirfRANS result (0.01419, PR #2680). Grad-clip=0.5 was just tested and FAILED (0.01446, +1.9% worse) — tighter clipping was counterproductive because 98.6-100% of batches were clipped, creating overfitting.

Grad-clip=1.5 fills the gap between the winning 1.0 and the looser 2.0 (being tested by winry). Since clip=0.5 was too tight, the optimal threshold may be ABOVE 1.0. At clip=1.5, fewer batches will be clipped than at 1.0 (where 91-98% were clipped), allowing more gradient signal through while still preventing the worst spikes.

## Instructions

Run the exact winning AirfRANS config with grad-clip=1.5:

```bash
cd target/icml2026 && python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 7e-4 --cosine-t-max 10 \
  --grad-clip 1.5 \
  --no-use-ema --enable-fourier --epochs 999 \
  --wandb-group airfrans-gradclip-sweep \
  --wandb-name "haku/airfrans-lr7e4-gradclip15"
```

Report:
1. Best val_primary/surface_mse and epoch
2. Grad-clip statistics: what fraction of batches are clipped?
3. Spike analysis: compare peak val_primary values at cosine LR peaks vs baseline

## Baseline

- **Primary metric:** val_primary/surface_mse = **0.01419** (PR #2680, lr=7e-4+grad-clip=1.0)
- **Grad-clip sweep so far:** clip=0.3 (roy #2717), clip=0.5 (FAILED, 0.01446), clip=1.0 (WINNER, 0.01419), clip=1.5 (THIS), clip=2.0 (winry #2718)
- **W&B baseline run:** 48ldl625
- **Reproduce baseline:** `cd target/icml2026 && python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 10 --grad-clip 1.0 --no-use-ema --enable-fourier --epochs 999`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT haku:

## Results

### Key Metrics

| Metric | Baseline (clip=1.0) | This Run (clip=1.5) | Change |
|--------|---------------------|----------------------|--------|
| **val_primary/surface_mse** | **0.01419** | **0.00935** (epoch 40) | **-34.1%** |
| test_primary/surface_mse | — | 0.01915 | — |
| full_test/volume_mse | — | 0.07964 | — |

### Grad-Clip Statistics

Fraction of batches clipped per epoch (360 train steps/epoch):

| Epoch | Clipped / Total | % Clipped | Grad Norm Mean |
|-------|----------------|-----------|----------------|
| 1 | 351/360 | 97.5% | 11.7 |
| 10 | 331/360 | 91.9% | 21.6 |
| 20 | 320/360 | 88.9% | 21.0 |
| 30 | 295/360 | 81.9% | 13.1 |
| 40 | 263/360 | 73.1% | 9.9 |

At clip=1.5, the clipping fraction starts high (~97%) and decreases to ~73% by epoch 40 as the model stabilizes. Compare to baseline clip=1.0 where 91-98% of batches were clipped throughout. The looser threshold allows significantly more gradient signal through, especially in later training.

### Spike Analysis (Cosine LR Peaks)

| Cycle Trough | Epoch | val_primary/surface_mse |
|-------------|-------|------------------------|
| Cycle 1 | 10 | 0.2270 |
| Cycle 2 | 20 | 0.0705 |
| Cycle 3 | 30 | 0.0774 |
| Cycle 4 | 40 | **0.00935** |

Peak-of-cycle spikes (worst val during each LR peak):
- Cycle 1 peak (ep 5): 0.338
- Cycle 2 peak (ep 13): 0.221
- Cycle 3 peak (ep 22): 0.148
- Cycle 4 peak (ep 32): 0.033

Spikes diminish across cycles — the model becomes more robust to LR restarts as training progresses.

### Command

```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 7e-4 --cosine-t-max 10 \
  --grad-clip 1.5 \
  --no-use-ema --enable-fourier --epochs 999 \
  --agent haku \
  --wandb-group airfrans-gradclip-sweep \
  --wandb-name "haku/airfrans-lr7e4-gradclip15"
```

- **W&B run:** `7bdiqnmi`
- **Peak memory:** Not logged (GPU 0, H100 97GB — no OOM issues)
- **Training stopped:** Epoch 40 (30-min timeout)

### What Happened

**This is a clear win.** Grad-clip=1.5 beats the baseline (clip=1.0) by 34.1% on val_primary/surface_mse (0.00935 vs 0.01419).

The key insight is that clip=1.5 allows more gradient signal through — by epoch 40 only 73% of batches are clipped vs 91-98% at clip=1.0. This lets the optimizer make larger, more informative updates while still preventing the worst gradient spikes. The grad norm mean also decreases from ~22 to ~10 over training, showing natural stabilization.

The cosine trough at epoch 40 hit the best value, suggesting the model could improve further with more training time. The convergence trajectory is strongly downward across cycles.

### Suggested Follow-ups

1. **Longer training**: The model was clearly still improving at epoch 40. Running with a longer timeout could yield even better results.
2. **Grad-clip=1.25 or 1.75**: Now that we know the optimal is above 1.0, a finer sweep around 1.5 could further optimize.
3. **Combine with other winners**: If clip=1.5 proves robust across seeds, combine with other winning hyperparameters.

**Comment by tcapelle** [comment]
ADVISOR: Merging — **0.00935** val_primary/surface_mse is a staggering 26.4% improvement over the 0.01271 baseline. Grad-clip=1.5 at T_max=10 found a dramatically deeper basin. This closes the gap to external target from 3x to ~2.2x. The finding that clip=1.5 > clip=1.0 > clip=0.5 suggests the gradient norm distribution needs room to operate — moderate clipping (1.5) prevents catastrophic spikes while allowing enough gradient magnitude for effective learning. Outstanding work haku!
